tiledb/sm/query/query_condition.cc

```diff
@@ -0,0 +1,414 @@
+  switch (attribute->type()) {
+    case Datatype::INT8:
+      return apply_clause<int8_t>(
+          clause, stride, fill_value, result_cell_slabs, out_result_cell_slabs);
+    case Datatype::UINT8:
+      return apply_clause<uint8_t>(
+          clause, stride, fill_value, result_cell_slabs, out_result_cell_slabs);
+    case Datatype::INT16:
+      return apply_clause<int16_t>(
+          clause, stride, fill_value, result_cell_slabs, out_result_cell_slabs);
+    case Datatype::UINT16:
+      return apply_clause<uint16_t>(
+          clause, stride, fill_value, result_cell_slabs, out_result_cell_slabs);
+    case Datatype::INT32:
+      return apply_clause<int32_t>(
+          clause, stride, fill_value, result_cell_slabs, out_result_cell_slabs);
+    case Datatype::UINT32:
+      return apply_clause<uint32_t>(
+          clause, stride, fill_value, result_cell_slabs, out_result_cell_slabs);
+    case Datatype::INT64:
+      return apply_clause<int64_t>(
+          clause, stride, fill_value, result_cell_slabs, out_result_cell_slabs);
+    case Datatype::UINT64:
+      return apply_clause<uint64_t>(
+          clause, stride, fill_value, result_cell_slabs, out_result_cell_slabs);
+    case Datatype::FLOAT32:
+      return apply_clause<float>(
+          clause, stride, fill_value, result_cell_slabs, out_result_cell_slabs);
+    case Datatype::FLOAT64:
+      return apply_clause<double>(
+          clause, stride, fill_value, result_cell_slabs, out_result_cell_slabs);
+    case Datatype::STRING_ASCII:
+      return apply_clause<char*>(
+          clause, stride, fill_value, result_cell_slabs, out_result_cell_slabs);
+    case Datatype::CHAR:
+      return apply_clause<char>(
+          clause, stride, fill_value, result_cell_slabs, out_result_cell_slabs);
+    case Datatype::DATETIME_YEAR:
+    case Datatype::DATETIME_MONTH:
+    case Datatype::DATETIME_WEEK:
+    case Datatype::DATETIME_DAY:
+    case Datatype::DATETIME_HR:
+    case Datatype::DATETIME_MIN:
+    case Datatype::DATETIME_SEC:
+    case Datatype::DATETIME_MS:
+    case Datatype::DATETIME_US:
+    case Datatype::DATETIME_NS:
+    case Datatype::DATETIME_PS:
+    case Datatype::DATETIME_FS:
+    case Datatype::DATETIME_AS:
+      return apply_clause<int64_t>(
+          clause, stride, fill_value, result_cell_slabs, out_result_cell_slabs);
+    case Datatype::ANY:
+    case Datatype::STRING_UTF8:
+    case Datatype::STRING_UTF16:
+    case Datatype::STRING_UTF32:
+    case Datatype::STRING_UCS2:
+    case Datatype::STRING_UCS4:
+    default:
+      return Status::QueryConditionError(
+          "Cannot perform query comparison; Unsupported query "
+          "conditional type on " +
+          clause.field_name_);
+  }
+
+  return Status::Ok();
+}
+
+Status QueryCondition::apply(
+    const ArraySchema* const array_schema,
+    std::vector<ResultCellSlab>* const result_cell_slabs,
+    const uint64_t stride) const {
+  if (clauses_.empty()) {
+    return Status::Ok();
+  }
+
```

Comment:
I'll preemptively draw attention to this for-loop. This serially iterates through each clause, mutating `result_cell_slabs` so that each slab only contains cells that fit the clause criteria. I have decided to leave this serial in anticipation of a refactor.

However -- it would not be much work to parallelize `apply_clause` for partitions of `result_cell_slabs`.